### ISSUE TITLE
Rollback

### ISSUE THREAD

**Issue by gaynetdinov** [boldpoker/edeliver#96]
Hello. 

I was wondering if there is a way to rollback a release/upgrade deploy. 
For example if after successful `upgrade` turned out that application got broken and I need very quickly rollback `upgrade`, i.e. start previously deployed release/upgrade.
Or another example: I just deployed a release and ran `restart` command and application got broken or something went wrong and I need to stop new release and start the old one. 

Is it possible to do something like that? 

Thanks!

**Comment by bharendt** [comment]
At the moment there is no builtin edeliver command to downgrade a release to a previous version.
 
But as long as downgrade instructions for the old version exist in the `relup` file of the deployed release, you can always downgrade using the `:release_handler.install_release('old-version')` function. But in case of an emergency you can do something like this:

```bash
export APP=my_app
export SSH_TIMEOUT=3
source .deliver/config 
source deps/edeliver/libexec/output
source deps/edeliver/libexec/erlang
source deps/edeliver/strategies/erlang-node-execute
# use staging or production here
NODE_ENVIRONMENT="staging" require_node_config 
# display the available old releases
APP="$APP" execute_node_command "rpc release_handler which_releases"
export DOWNGRADE_VERSION="old-release-version"
# downgrade to previous version
APP="$APP" execute_node_command "rpc release_handler install_release '[\"$DOWNGRADE_VERSION\"].'"
```

I think it would make sense to add a `downgrade` command to edeliver, but I think only if also the following features would be included:

- ability to show the available releases you can downgrade to
- autodetect the previous version to downgrade to
- allow to pass a version to downgrade to
- make the old release permanent (= shold boot on restart)

For the other case, that you deployed and restarted the release and want to restart the old one, you must deploy and restart the old release, except there would be an option from exrm to activate the old release. exrm sym-links the start-, boot- and config-files and it would be required to link them back to the old version.

**Comment by geofflane** [comment]
Just ran into a need for this. :) 👍

**Comment by gaynetdinov** [comment]
@bharendt thanks again for the explanation!

**Comment by bharendt** [comment]
I also think this would be a useful feature, even if it might not be necessary so often.

 Being able to upgrade a running server without restarting it e.g without disconnecting clients, disconnecting other servers (e.g. a database), without losing state etc., is one of the very nice and often overlooked features of the erlang vm. 

And even being able to downgrade to the last version if the upgrade didn't fulfill the expectations or just doesn't work is even better. In a reliable production system you might actually always want to prepare also the downgrade (and also test it), so I think it makes quite much sense to add a `downgrade` command. And it should not require so much new edeliver code.

Downgrading to an arbitrary version maybe does not make so much sense and also would require to provide downgrade instructions in each new release to any old version, so I think this does not need to be implemented while the last point, making it permanent should work out of the box because it is acutally the same as to make a new release the permanent one.

**Comment by philipgiuliani** [comment]
This feature would have been awesome for me just now :) +1

**Comment by ajanes** [comment]
Yes, would be a useful feature, agree that downgrading to an arbitrary version does not make much sense on a production level.

**Comment by fairbairn** [comment]
+1

Just the ability to start the older release version would be a life saver.  Right now we have to deploy the same older release version again to start-deploy.  Is there a way to just reference the older version already uploaded to the server and switch to that one instead of the newer one?
